### PR TITLE
Remove overlapping exclusive accesses in encoders

### PR DESCRIPTION
**Pull request by itaiferber**
**What's in this pull request?**
{JSON,Plist}{Encoder,Decoder} had overlapping accesses which were supposed to be mutually exclusive in their `with(pushedKey:)` methods.

Removes those methods and the overlapping accesses.

### CODE REVIEW

**Comment by itaiferber** [comment]
@swift-ci Please test macOS

**Comment by itaiferber** [comment]
@swift-ci Please smoke test Linux

**Review by parkera — APPROVED** [review]

File: stdlib/public/SDK/Foundation/JSONEncoder.swift [hunk 1]
```diff
@@ -1374,287 +1342,305 @@ fileprivate struct _JSONUnkeyedDecodingContainer : UnkeyedDecodingContainer {
...
-            guard let decoded = try self.decoder.unbox(self.container[self.currentIndex], as: Int.self) else {
-                throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: self.decoder.codingPath + [_JSONKey(index: self.currentIndex)], debugDescription: "Expected \(type) but found null instead."))
-            }
+        self.decoder.codingPath.append(_JSONKey(index: self.currentIndex))
+        defer { self.decoder.codingPath.removeLast() }
 
-            self.currentIndex += 1
-            return decoded
+        guard let decoded = try self.decoder.unbox(self.container[self.currentIndex], as: Int.self) else {
+            throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: self.decoder.codingPath + [_JSONKey(index: self.currentIndex)], debugDescription: "Expected \(type) but found null instead."))
```
Comment (norio-nomura) [inline stdlib/public/SDK/Foundation/JSONEncoder.swift]: It looks `DecodingError.valueNotFound` has duplicated `_JSONKey(index: self.currentIndex)` in `codingPath`.
Reproducing code:
```swift
import Foundation

struct A: Encodable {
    var field1: String?

    public func encode(to encoder: Encoder) throws {
        var container = encoder.unkeyedContainer()
        try container.encode(field1)
    }
}

struct B: Decodable {
    var field1: Int = 0

    public init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()
        field1 = try container.decode(Int.self)
    }
}

do {
    let b = try JSONDecoder().decode(B.self, from: JSONEncoder().encode(A()))
    print(b)
} catch {
    print(error)
}
```
> valueNotFound(Swift.Int, Swift.DecodingError.Context(codingPath: [Foundation.(_JSONKey in _12768CA107A31EF2DCE034FD75B541C9)(stringValue: "Index 0", intValue: Optional(0)), Foundation.(_JSONKey in _12768CA107A31EF2DCE034FD75B541C9)(stringValue: "Index 0", intValue: Optional(0))], debugDescription: "Expected Int but found null instead.", underlyingError: nil))

Comment (itaiferber) [inline stdlib/public/SDK/Foundation/JSONEncoder.swift]: @norio-nomura Thanks for letting me know! I'll look into it

Comment (itaiferber) [inline stdlib/public/SDK/Foundation/JSONEncoder.swift]: @norio-nomura What compiler version/Xcode version are you using? I see a different error on my end.

Comment (itaiferber) [inline stdlib/public/SDK/Foundation/JSONEncoder.swift]: Sorry, nevermind — I was on the wrong Swift version myself. I get the same result as you do. I'll have to hunt down where we append the key to the coding path twice

**Comment by jrose-apple** [comment]
If you wanted to keep this `with` thing you could have the callback take an `inout Encoder` and then use `$0` inside. But this works too.

File: stdlib/public/SDK/Foundation/JSONEncoder.swift (hunk 1 above)
Comment (norio-nomura) [inline stdlib/public/SDK/Foundation/JSONEncoder.swift]: The key is appended at the line that I commented on.
Same codes exist on each `decode(_:)` methods of `_JSONUnkeyedDecodingContainer` and `_PlistUnkeyedDecodingContainer`.

**Pull request merged**